apps/api/src/app/workflows-v2/usecases/generate-preview/generate-preview.usecase.ts

```diff
@@ -17,74 +18,100 @@ import {
 
 const LOG_CONTEXT = 'GeneratePreviewUsecase';
 
+type DestructuredControlValues = {
+  tiptapControlValues: { emailEditor?: string | null; body?: string | null } | null;
+  // this is the remaining control values after the tiptap control is extracted
+  simpleControlValues: Record<string, unknown>;
+};
+
+type ProcessedControlResult = {
+  controlValues: Record<string, unknown>;
+  variablesExample: Record<string, unknown> | null;
+};
+
 @Injectable()
 export class GeneratePreviewUsecase {
   constructor(
-    private legacyPreviewStepUseCase: PreviewStep,
+    private previewStepUsecase: PreviewStep,
     private buildStepDataUsecase: BuildStepDataUsecase,
     private getWorkflowByIdsUseCase: GetWorkflowByIdsUseCase,
     private readonly logger: PinoLogger,
-    private prepareAndValidateContentUsecase: PrepareAndValidateContentUsecase,
-    private buildPayloadSchema: BuildPayloadSchema
+    private buildPayloadSchema: BuildPayloadSchema,
+    private prepareAndValidateContentUsecase: PrepareAndValidateContentUsecase
   ) {}
 
   @InstrumentUsecase()
   async execute(command: GeneratePreviewCommand): Promise<GeneratePreviewResponseDto> {
     try {
-      const { previewPayload: commandVariablesExample, controlValues: commandControlValues } =
-        command.generatePreviewRequestDto;
-      const stepData = await this.getStepData(command);
-      const controlValues = commandControlValues || stepData.controls.values || {};
-      const workflow = await this.findWorkflow(command);
-      const payloadSchema = await this.buildPayloadSchema.execute(
-        BuildPayloadSchemaCommand.create({
-          environmentId: command.user.environmentId,
-          organizationId: command.user.organizationId,
-          userId: command.user._id,
-          workflowId: command.workflowIdOrInternalId,
-          controlValues,
-        })
-      );
+      const {
+        stepData,
+        controlValues: initialControlValues,
+        variableSchema,
+        workflow,
+      } = await this.initializePreviewContext(command);
 
-      const variableSchema = this.buildVariablesSchema(stepData.variables, payloadSchema);
-      const preparedAndValidatedContent = await this.prepareAndValidateContentUsecase.execute({
-        user: command.user,
-        previewPayloadFromDto: commandVariablesExample,
-        controlValues,
-        controlDataSchema: stepData.controls.dataSchema || {},
+      const sanitizedValidatedControls = sanitizePreviewControlValues(initialControlValues, stepData.type);
+
+      if (!sanitizedValidatedControls) {
+        throw new Error(
+          // eslint-disable-next-line max-len
+          'Control values normalization failed: The normalizeControlValues function requires maintenance to sanitize the provided type or data structure correctly'
+        );
+      }
+
+      const destructuredControlValues = this.destructureControlValues(sanitizedValidatedControls);
```

Comment:
@djabarovgeorge I think we can simplify this part even more. There is no need to use intermediate structures such as destructuredControlValues that group variables into two categories (tiptap and simple). Alternatively, we can use a single function that parses the controls one after the other and checks whether the control is tiptap or simple.

That function will return the final previewData object at once without doing a final merge. 

Effectively, it's a `reduce` function on the step controls array that returns the preview Data at once.

Let me know if this makes sense.